modules/rational-autocompile-el.el

```diff
@@ -0,0 +1,154 @@
+;;;; rational-autocompile-el.el --- Autocompile emacs-lisp code  -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2022
+;; SPDX-License-Identifier: MIT
+
+;; Author: System Crafters Community
+
+
+;;; Commentary:
+
+;; This module defines variables, modules and hooks to enable
+;; automatic compilation of emacs-lisp source files.
+
+;; This could be activated for the `modules' subdirectory under your
+;; `user-emacs-directory', your files under `rational-config-path', or
+;; all files with mode `emacs-lisp-mode' after saving the file.
+
+;; It also chooses to compile to byte-code or native code depending if
+;; the feature is available on the current build.
+
+
+;;; Code:
+
+;;;; Variables:
+;; We define a variable to enable/disable modules autocompilation.
+(defvar rational-autocompile-el-modules t
+  "When non-nil, autocompile emacs lisp sources for the modules on startup.")
+
+(defvar rational-autocompile-el-user-configuration t
+  "When non-nil, autocompile emacs lisp sources for the modules on startup.")
+
+(defvar rational-autocompile-el-init-files nil
+  "When non-nil, autocompile init files on startup.")
+
+(defvar rational-autocompile-el-on-save nil
+  "When non-nil, autocompile the file after saving it.")
+
+(defvar rational-autocompile-el-extra-directories-list nil
+  "List of extra directories to autocompile.")
+
+(defvar rational-autocompile-el-init-files-list
+  '("early-init.el" "init.el")
+  "List of initialization file names.")
+
+;;;; Functions:
+;; A function to compile a specific file
+(defun rational-autocompile-el-file (f)
+  "Compiles (native or byte-code) the file F.
+
+F could be a single file or a list of files.
+
+If F is a directory or contains a directory, the content of that
+directory will be compiled, but not it's subdirectories."
+  (setq f (flatten-list (list f)))
+  (message "Compiling file(s): %s" f)
+  (if (featurep 'native-compile)
+      (native-compile-async f)
+    (dolist (source f)
+      (when (file-exists-p source)
+        (if (file-directory-p source)
+            (byte-recompile-directory source 0)
+          (byte-recompile-file source nil 0))))))
+
+;; A function to compile the buffer's file
+(defun rational-autocompile-el-buffer (&optional b)
+  "Compiles (native or byte-code) the file of buffer B."
+  (when b
+    (unless (bufferp b)
+      return nil))
```

Comment:
This should be `cl-return` instead of return.

Reply:
Oh, right! It also needs parens.

Reply:
Done!